x86_64/src/trap.rs

```diff
@@ -149,6 +140,20 @@ pub unsafe extern "C" fn trap() -> ! {
             movq %rcx, 2*8(%rsp);
             movq %rbx, 1*8(%rsp);
             movq %rax, 0*8(%rsp);
+            // Save the x86 segmentation registers.  Uses %rdi
+            // as a scratch register, so we do this after we've
+            // saved the GP registers.  Note that we clear %rdi
+            // once, and then just move into the low bits and
+            // copy to the stack.
+            xorl %edi, %edi
```

Comment:
xorq %rdi, %rdi

clearer to the reader.

Reply:
Hmm. The `xorl` encoding is just a tad denser than `xorq`, and the Intel optimization manual recommends it when applicable. I added a comment to clarify though; let me know what you think?

Reply:
Actually...scratch that. I just noticed that you can do a 64-bit `movq` from a 16-bit segmentation register into a 64-bit register, and that will be zero-extended, so no need for the `xor` shenanigans at all.

Reply:
Actually...scratch the scratch. I've tried to explain this in comments, but `movq` uses _more_ instruction bytes than what had been there. However, there's an encoding where good 'ol `xorl` will zero-extend the source (segmentation) register to the destination, and when combined with the CPU extending a 32-bit write into a 64-bit register, this has the desired effect of zeroing the upper 48 bits of %rdi. So the xor is gone, and I think we've hit approximately the minimal encoding.

Reply:
Lol.  So I had seen your first reply only, which left me wondering where you were going with this.  It still feels a bit early to me to be optimizing the assembly here for anything other than readability, but I think you found a good tradeoff 😅